typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1361 @@
+     shared and a child uniquely is obviously bad. Howerver, it might be the
+     result of "par"ing multiple path: `S` par `N -> U`, which is valid.
+
+     INVARIANT: children >= parent. For example, having a shared child under a
+     unique parent is nonsense. The invariant is preserved because Usage.par and
+     Usage.seq above is monotone, and Usage_tree.par and Usage_tree.seq here is
+     node-wise. *)
+
+  let rec print_children ppf children =
+    Projection.Map.iter
+      (fun proj child ->
+        Format.fprintf ppf "%a = %a," Projection.print proj print child)
+      children
+
+  and print ppf t =
+    Format.fprintf ppf "%a {%a}" Usage.print t.usage print_children t.children
+
+  module Path = struct
+    type t = Projection.t list
+
+    let child (p : t) (a : Projection.t) : t = p @ [ a ]
+    let root : t = []
+    let _print ppf = Format.pp_print_list Projection.print ppf
+  end
+
+  let leaf usage = { usage; children = Projection.Map.empty }
+  let _unused = leaf Usage.Unused
+
+  (** f must be monotone *)
+  let map_with_projs projs f t =
+    let rec loop projs t =
+      let usage = f projs t.usage in
+      let children =
+        Projection.Map.mapi (fun proj t -> loop (proj :: projs) t) t.children
+      in
+      { usage; children }
+    in
+    loop projs t
+
+  let map f t = map_with_projs [] f t
+
+  let rec map2 f t0 t1 =
+    let usage = f Self t0.usage t1.usage in
+    let children =
+      Projection.Map.merge
+        (fun proj c0 c1 ->
+          match (c0, c1) with
+          | None, None -> assert false
+          | None, Some c1 ->
+              Some
+                (map_with_projs [ proj ]
+                   (fun projs r -> f (Ancestor projs) t0.usage r)
+                   c1)
+          | Some c0, None ->
+              Some
+                (map_with_projs [ proj ]
+                   (fun projs l -> f (Descendant projs) l t1.usage)
+                   c0)
+          | Some c0, Some c1 -> Some (map2 f c0 c1))
+        t0.children t1.children
+    in
+    { usage; children }
+
+  let par t0 t1 = map2 (fun _ -> Usage.par) t0 t1
+
+  let seq t0 t1 =
+    map2
+      (fun rel t0 t1 ->
+        try Usage.seq t0 t1
+        with Error (MultiUse { first; second; axis; first_or_second; _ }) ->
+          raise
+            (Error
+               (MultiUse
+                  {
+                    first;
+                    second;
+                    axis;
+                    first_or_second;
+                    first_is_of_second = Some rel;
+                  })))
+      t0 t1
+
+  let rec singleton path leaf =
+    match path with
+    | [] -> { usage = leaf; children = Projection.Map.empty }
+    | proj :: path ->
+        {
+          usage = Unused;
+          children = Projection.Map.singleton proj (singleton path leaf);
+        }
+end
+
+(** Lift Usage_tree to forest *)
+module Usage_forest : sig
+  module Path : sig
```

Comment:
I think this interface is actually easier to understand if it includes the `Root_id` (which I would just call `Root`). Then expose `Root.fresh`, `Root.fresh_of_ident` and `Path.root`, instead of `Path.fresh_root` and `Path.fresh_root_of_ident`.

Reply:
Hmm then we need to expose `UF.Path.t` as the product of `Root` and `Usage_tree.Path.t`? But I feel there is no much practical benefit in that?

Reply:
I agree there is space for abstracting "list of `UF.Paths.t`" to its own type, but I'm not sure about having this inside `module UF`, as I really like the idea that any `UF.t` can be thought as a mapping from `UF.Path.t` to `Usage`.

Reply:
A single `Path.t` represents some value in our notion of "heap". It's natural to end up moving to the powerset of such a notion once there is some form of control flow involved. In our case this comes from or-patterns, but we'd get the same thing if we tried to make the analysis more precise on expressions (e.g. tracking `let x = if p then y else z in` accurately). It's no less algebraic than the rest of what is being defined here.

Reply:
We now have a `module Paths` outside of `Usage_forest` - could you take a look?